### PR TITLE
Convert workspace unit validator to accept multiple units

### LINKED ISSUE #20410: Make WorkspaceUnitValidator work for multiple units
Currently `WorkspaceUnitValidator` will only work for a single unit type. For #20409 it would be useful to check if the workspace has units of either TOF or d spacing.

### PR DESCRIPTION
**Pull request by samueljackson92**
This converts the `WorkspaceUnitValidator` to accept both a single unit and a vector of unit ids.

**To test:**
 - Unit tests are passing for existing behaviour
 - New units tests adequately cover the new behaviour.
 - Code review.

Fixes #20410 

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/inc/MantidAPI/WorkspaceUnitValidator.h [hunk 1]
```diff
@@ -33,7 +33,11 @@ namespace API {
 */
 class MANTID_API_DLL WorkspaceUnitValidator : public MatrixWorkspaceValidator {
```
Comment (NickDraper) [inline Framework/API/inc/MantidAPI/WorkspaceUnitValidator.h]: A more general design would have been to enhance the CompositeValidator, or make another one that allowed OR operations.

File: Framework/API/src/WorkspaceUnitValidator.cpp [hunk 2]
```diff
@@ -38,19 +53,39 @@ WorkspaceUnitValidator::checkValidity(const MatrixWorkspace_sptr &value) const {
...
-    if ((!unit) || (unit->unitID().compare(m_unitID))) {
-      return "The workspace must have units of " +
-             m_unitID; //+ "; its unit is: " + unit->caption();
-    } else
-      return "";
+    const auto matchesUnitID =
+        [&unit](const std::string &unitID) { return unit->unitID() == unitID; };
+
+    if (!unit ||
+        !std::any_of(m_unitIDs.cbegin(), m_unitIDs.cend(), matchesUnitID)) {
```
Comment (NickDraper) [inline Framework/API/src/WorkspaceUnitValidator.cpp]: I find NOT OR operations slightly unintuitive to see in the code clearly, a comment here might help.

**Review by NickDraper — CHANGES_REQUESTED** [review]
Don't worry too much about my comment on the design, just add the comment and this is good to go.

**Review by NickDraper — APPROVED** [review]

**Comment by NickDraper** [comment]
This is good now, and worth mentioning at the next mantid review

File: Framework/Kernel/inc/MantidKernel/CompositeValidator.h [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include <boost/make_shared.hpp>
 
 #include <list>
+#include <sstream>
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/CompositeValidator.h]: Is this header required? This gets included in a lot of other places so it's good to keep these core headers as small as possible.

Comment (samueljackson92) [reply Framework/Kernel/inc/MantidKernel/CompositeValidator.h]: No. Will move it.

Comment (martyngigg) [reply Framework/Kernel/inc/MantidKernel/CompositeValidator.h]: :+1:

**Pull request merged**